spacy/tokenizer.pyx

```diff
@@ -174,23 +211,124 @@ cdef class Tokenizer:
             yield self(text)
 
     def _flush_cache(self):
-        self._reset_cache([key for key in self._cache if not key in self._specials])
+        self._reset_cache([key for key in self._cache])
 
     def _reset_cache(self, keys):
         for k in keys:
+            cached = <_Cached*>self._cache.get(k)
             del self._cache[k]
-            if not k in self._specials:
-                cached = <_Cached*>self._cache.get(k)
-                if cached is not NULL:
-                    self.mem.free(cached)
+            if cached is not NULL:
+                self.mem.free(cached)
 
-    def _reset_specials(self):
+    def _flush_specials(self):
         for k in self._specials:
             cached = <_Cached*>self._specials.get(k)
             del self._specials[k]
             if cached is not NULL:
                 self.mem.free(cached)
 
+    cdef int _apply_special_cases(self, Doc doc):
```

Comment:
I guess this can be `void` return type?

Reply:
This is a good thing to highlight, as there's some subtleties to Cython signatures that are extremely non-obvious. It's why we often have `int` as the return type of functions and `except -1`. Maybe this is resolved in current versions of Cython, but my understanding is:

Cython uses the `except` value as the error code, so that you can raise errors within the function. Cython will call the Python error checking functions if the return value is equal to the exception value you set. If you have a C return type and don't specify a value as the exception value, when there's errors within the function, they're swallowed and you get that spam on your terminal about unraisable exceptions. Instead of specifying a particular value, you can also use `except *`. This forces Cython to check the Python error status after every function invocation. which introduces a bit of overhead.

If you use `void` as the return type, you have to use `except *`. So using `void` as a return type is slower --- which is extremely counter-intuitive, but there you go. The problem doesn't come up if you're making your function `nogil`, as a `nogil` function can't be raising any Python errors anyway (it can't call into any Python routines).

A similar problem occurs with the `bint` type. If you mark a function as `bint`, you don't really have a good value to use as an error. So `int` is usually better, so you can make it `except -1`.

Reply:
I think I used `int` because all the other functions I'd seen were `int`, but I didn't understand the details.

So this should stay `int` and I should add `except -1`? And `_filter_spans()` can stay `void ... nogil` and the cmp functions can stay `bint ... nogil`?

Reply:
Thanks for the explanation @honnibal ! I definitely did not realise all of that ...